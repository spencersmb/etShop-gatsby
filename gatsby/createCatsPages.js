const path = require("path")

module.exports = async ({ actions, graphql }) => {
  const { createPage } = actions
  const perPage = 10
  const allCats = {}
  const excludeItems = ["uncategorized"]
  let pageNumber = 1

  const fetchCatPages = async (variables) =>
    await graphql(GET_CATS, variables).then(async ({ data }) => {
      const {
        wpgraphql: {
          categories: {
            nodes
          }
        }
      } = data

      await Promise.all(nodes
        .filter(node => !excludeItems.includes(node.slug))
        .map((cat) => {
          const {
            supportQuestions: {
              pageInfo: { hasNextPage, endCursor, hasPreviousPage }
            }
          } = cat
          const questions = cat.supportQuestions.nodes
          const catPageTemplate = path.resolve(`./src/templates/supportCatListPage.tsx`)
          const catPagePath = `/support/category/${cat.slug}/page/${pageNumber}`
          // console.log("cat.supportQuestions.pageInfo", cat.supportQuestions.pageInfo)
          // console.log("slug", cat.slug)
          // console.log("ID", cat.id)
          // console.log("pageNumber", pageNumber)
          // console.log("hasNextPage", hasNextPage)
          // console.log("hasPrevPage", hasPreviousPage)

          // add item to be excluded if no next page
          if (!hasNextPage) {
            excludeItems.push(cat.slug)
          }
          const prevPages = allCats[cat.slug] && allCats[cat.slug].pages ? allCats[cat.slug].pages : {}
          // check if object exists first
          allCats[cat.slug] = {
            pages: {
              ...prevPages,
              [pageNumber]: {
                path: catPagePath,
                component: catPageTemplate,
                context: {
                  cat: cat.name,
                  catId: cat.id,
                  questions,
                  pageNumber,
                  hasNextPage,
                  hasPrevPage: hasPreviousPage
                }
              }
            }
          }

          if (hasNextPage) {
            pageNumber++
            return fetchCatPages({ first: perPage, after: endCursor })
          }
        }))
      return allCats
    })

  await fetchCatPages({ first: perPage, after: null }).then(catItems => {
    Object.keys(catItems).map((cat) => {
      const pages = Object.keys(catItems[cat].pages)
      pages.map(page => {
        console.log(`create Cat Page ${cat}`)
        console.log(`create Cat PageNumber ${catItems[cat].pages[page].context.pageNumber}`)
        createPage(catItems[cat].pages[page])
      })
    })
  })
}

const GET_CATS = `
  query GET_ALL_CATS($first:Int $after:String){
      wpgraphql{
        categories {
          nodes {
            id
            name
            slug
            supportQuestions(after: $after, first: $first) {
              pageInfo{
                  hasNextPage
                  hasPreviousPage
                  endCursor
              }
              nodes {
                  title
                  id
                  slug
                  excerpt
                  acfSupportQuestions{
                      popularity
                  }
              }
            }
          }
        }
      }
    }
  `
