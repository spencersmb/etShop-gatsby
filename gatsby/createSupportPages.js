const path = require(`path`)
module.exports = async ({ actions, graphql }) => {
  const supportQuery = `
  {
  wpgraphql {
    supportQuestions(where: {status: PUBLISH}, first: 100){
      nodes{
        date
        uri
        id
        slug
        title
        content
        featuredImage {
          altText
          caption
          sourceUrl
        }
      }
    }
  }
}
  `

  return new Promise((resolve, reject) => {
    graphql(supportQuery).then(results => {

      if (results.errors) {
        console.log(results.errors)
        reject(results.errors)
      }
      results.data.wpgraphql.supportQuestions.nodes.forEach(question => {
        actions.createPage({
          path: `/support/${question.slug}/`,
          component: path.resolve(`./src/templates/supportQuestion.tsx`),
          context: {
            // blogId: blog.id,
            ...question
          }
        })
        console.log(`Created page: /support/${question.slug}`)
      })
      resolve()
    })

  })
}
