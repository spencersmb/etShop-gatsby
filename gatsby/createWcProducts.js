const path = require("path")
const GET_PAGES = `
  query GET_WC_PAGES{
      allWcProduct{
        edges{
          node{
            slug
            name
          }
        }
      }
    }
  `
module.exports = async ({ actions, graphql }) => {
  const { createPage } = actions
  const allPages = []
  const fetchPages = async () =>
    await graphql(GET_PAGES).then(({ data }) => {
      const {
        allWcProduct: {
          edges
        }
      } = data
      edges.map(page => {
        allPages.push(page.node)
      })
      return allPages
    })

  await fetchPages().then(allPages => {
    allPages.map(page => {
      console.log(`create page: ${page.slug}`)
      createPage({
        path: `/products/${page.slug}`,
        component: path.resolve(`./src/templates/productPage.tsx`),
        context: {
          slug: page.slug
        }
      })
    })
  })

  // return new Promise((resolve, reject) => {
  //   graphql(pageQuery).then(results => {
  //
  //     if (results.errors) {
  //       console.log(results.errors)
  //       reject(results.errors)
  //     }
  //     results.data.allWcProduct.edges.forEach(({ node }) => {
  //       createPage({
  //         path: `/products/${node.slug}`,
  //         component: path.resolve(`./src/components/products/productDetailPage.tsx`),
  //         context: {
  //           slug: node.slug
  //         }
  //       })
  //       console.log(`Created page: /products/${node.slug}`)
  //     })
  //   })
  //   resolve()
  // })
}
