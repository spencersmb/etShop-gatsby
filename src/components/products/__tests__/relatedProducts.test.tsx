import RelatedProducts from '@components/products/modules/relatedProducts'
import RelatedProduct from '@components/products/modules/relatedProductWrapper'
import React from 'react'
import { render } from 'react-testing-library'
import { StaticQuery } from 'gatsby'
import renderer from 'react-test-renderer'

const defaultProps = [
	'watercolor-texture-kit-vol-1',
	'watercolor-texture-kit-vol-2'
]

beforeEach(() => {
	// @ts-ignore
	StaticQuery.mockImplementationOnce((mock: any) =>
		mock.render({
			site: {
				siteMetadata: {
					title: `Default`
				}
			},
			allWcProduct: {
				edges: [
					{
						node: {
							name: 'Dreamy Ink Textures',
							id: '46e80e85-da3a-5a1e-9ef3-bd46e38d1dd5',
							slug: 'watercolor-texture-kit-vol-1',
							sub_header: 'Alcohol Ink Texture Pack',
							price: '15',
							featuredImage: {
								alt: 'Dreamy Ink Texture Pack',
								localFile: {
									childImageSharp: {
										fluid: {
											aspectRatio: 12,
											srcSet: 'srcSet',
											sizes: '',
											src: '/static/dreamy-textures-800x500.jpg'
										}
									}
								}
							}
						}
					},
					{
						node: {
							name: 'Dreamy Ink Textures 2',
							id: '46e80e85-da3a-5a1e-6666-bd46e38d1dd5',
							slug: 'watercolor-texture-kit-vol-2',
							sub_header: 'Alcohol Ink Texture Pack',
							price: '15',
							featuredImage: {
								alt: 'Dreamy Ink Texture Pack',
								localFile: {
									childImageSharp: {
										fluid: {
											aspectRatio: 12,
											srcSet: 'srcSet',
											sizes: '',
											src: '/static/dreamy-textures-800x500.jpg'
										}
									}
								}
							}
						}
					}
				]
			}
		})
	)
})

const data = {
	site: {
		siteMetadata: {
			title: `Default`
		}
	},
	allWcProduct: {
		edges: [
			{
				node: {
					name: 'Dreamy Ink Textures',
					id: '46e80e85-da3a-5a1e-9ef3-bd46e38d1dd5',
					slug: 'watercolor-texture-kit-vol-1',
					sub_header: 'Alcohol Ink Texture Pack',
					price: '15',
					featuredImage: {
						alt: 'Dreamy Ink Texture Pack',
						localFile: {
							childImageSharp: {
								fluid: {
									aspectRatio: 12,
									srcSet: 'srcSet',
									sizes: '',
									src: '/static/dreamy-textures-800x500.jpg'
								}
							}
						}
					}
				}
			},
			{
				node: {
					name: 'Dreamy Ink Textures 2',
					id: '46e80e85-da3a-5a1e-6666-bd46e38d1dd5',
					slug: 'watercolor-texture-kit-vol-2',
					sub_header: 'Alcohol Ink Texture Pack',
					price: '15',
					featuredImage: {
						alt: 'Dreamy Ink Texture Pack',
						localFile: {
							childImageSharp: {
								fluid: {
									aspectRatio: 12,
									srcSet: 'srcSet',
									sizes: '',
									src: '/static/dreamy-textures-800x500.jpg'
								}
							}
						}
					}
				}
			}
		]
	}
}
describe('Related Products', () => {
	it('Related Products renders correctly', () => {
		const tree = renderer
			.create(
				<RelatedProducts products={defaultProps} data={data}/>
			)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('Related Product renders correctly', () => {
		const tree = renderer
			.create(
				<RelatedProduct slug={defaultProps[0]} data={data}/>
			)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('Should render number of related Products', () => {
		const modalRender = render(<RelatedProducts products={defaultProps} data={data}/>)
		expect(modalRender.getByTestId('productsList').children.length).toEqual(2)
	})

	it('Should render number of related Products', () => {
		const modalRender = render(<RelatedProduct slug={defaultProps[0]} data={data}/>)
		expect(modalRender.getByTestId('listItem').children.length).toEqual(1)
	})
})
