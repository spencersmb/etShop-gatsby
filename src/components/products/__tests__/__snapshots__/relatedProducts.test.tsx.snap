// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related Products Related Product renders correctly 1`] = `
<div
  className="sc-bdVaJa hKSJDv"
  data-testid="listItem"
>
  <a
    href="/products/watercolor-texture-kit-vol-1"
  >
    <div
      className="sc-bwzfXH djTlgz"
    >
      <div
        className="ListItemTop sc-bxivhb cXSKfV"
      >
        <div
          className=" gatsby-image-wrapper"
          style={
            Object {
              "overflow": "hidden",
              "position": "relative",
            }
          }
        >
          <div
            style={
              Object {
                "paddingBottom": "8.333333333333334%",
                "width": "100%",
              }
            }
          />
          <picture>
            <img
              alt="Dreamy Ink Texture Pack"
              onLoad={[Function]}
              sizes=""
              src="/static/dreamy-textures-800x500.jpg"
              srcSet="srcSet"
              style={
                Object {
                  "height": "100%",
                  "left": 0,
                  "objectFit": "cover",
                  "objectPosition": "center",
                  "opacity": 0,
                  "position": "absolute",
                  "top": 0,
                  "transition": "opacity 0.5s",
                  "width": "100%",
                }
              }
            />
          </picture>
          <noscript
            dangerouslySetInnerHTML={
              Object {
                "__html": "<picture><img srcset=\\"srcSet\\" src=\\"/static/dreamy-textures-800x500.jpg\\" alt=\\"Dreamy Ink Texture Pack\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
              }
            }
          />
        </div>
        <div
          className="sc-htpNat kicMsp"
        >
          <h2>
            Dreamy Ink Textures
          </h2>
          <p>
            Alcohol Ink Texture Pack
          </p>
        </div>
      </div>
      <div
        className="ListItemFooter sc-ifAKCX gYZfIE"
      >
        <span />
        <p>
          $
          15
        </p>
        <test-file-stub />
      </div>
    </div>
  </a>
</div>
`;

exports[`Related Products Related Products renders correctly 1`] = `
<div
  className="sc-EHOje sc-dnqmqq iAnCZo"
>
  <div
    className="sc-htoDjs hrMTXy"
  >
    <h5>
      Related Products
    </h5>
    <div>
      <a
        href="/"
      >
        View All products
      </a>
    </div>
  </div>
  <div
    className="sc-iwsKbI bIrZoZ"
    data-testid="productsList"
  >
    <div
      className="sc-bdVaJa hKSJDv"
      data-testid="listItem"
    >
      <a
        href="/products/watercolor-texture-kit-vol-1"
      >
        <div
          className="sc-bwzfXH djTlgz"
        >
          <div
            className="ListItemTop sc-bxivhb cXSKfV"
          >
            <div
              className=" gatsby-image-wrapper"
              style={
                Object {
                  "overflow": "hidden",
                  "position": "relative",
                }
              }
            >
              <div
                style={
                  Object {
                    "paddingBottom": "8.333333333333334%",
                    "width": "100%",
                  }
                }
              />
              <picture>
                <img
                  alt="Dreamy Ink Texture Pack"
                  onLoad={[Function]}
                  sizes=""
                  src="/static/dreamy-textures-800x500.jpg"
                  srcSet="srcSet"
                  style={
                    Object {
                      "height": "100%",
                      "left": 0,
                      "objectFit": "cover",
                      "objectPosition": "center",
                      "opacity": 0,
                      "position": "absolute",
                      "top": 0,
                      "transition": "opacity 0.5s",
                      "width": "100%",
                    }
                  }
                />
              </picture>
              <noscript
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "<picture><img srcset=\\"srcSet\\" src=\\"/static/dreamy-textures-800x500.jpg\\" alt=\\"Dreamy Ink Texture Pack\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
                  }
                }
              />
            </div>
            <div
              className="sc-htpNat kicMsp"
            >
              <h2>
                Dreamy Ink Textures
              </h2>
              <p>
                Alcohol Ink Texture Pack
              </p>
            </div>
          </div>
          <div
            className="ListItemFooter sc-ifAKCX gYZfIE"
          >
            <span />
            <p>
              $
              15
            </p>
            <test-file-stub />
          </div>
        </div>
      </a>
    </div>
    <div
      className="sc-bdVaJa hKSJDv"
      data-testid="listItem"
    >
      <a
        href="/products/watercolor-texture-kit-vol-2"
      >
        <div
          className="sc-bwzfXH djTlgz"
        >
          <div
            className="ListItemTop sc-bxivhb cXSKfV"
          >
            <div
              className=" gatsby-image-wrapper"
              style={
                Object {
                  "overflow": "hidden",
                  "position": "relative",
                }
              }
            >
              <div
                style={
                  Object {
                    "paddingBottom": "8.333333333333334%",
                    "width": "100%",
                  }
                }
              />
              <picture>
                <img
                  alt="Dreamy Ink Texture Pack"
                  onLoad={[Function]}
                  sizes=""
                  src="/static/dreamy-textures-800x500.jpg"
                  srcSet="srcSet"
                  style={
                    Object {
                      "height": "100%",
                      "left": 0,
                      "objectFit": "cover",
                      "objectPosition": "center",
                      "opacity": 0,
                      "position": "absolute",
                      "top": 0,
                      "transition": "opacity 0.5s",
                      "width": "100%",
                    }
                  }
                />
              </picture>
              <noscript
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "<picture><img srcset=\\"srcSet\\" src=\\"/static/dreamy-textures-800x500.jpg\\" alt=\\"Dreamy Ink Texture Pack\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
                  }
                }
              />
            </div>
            <div
              className="sc-htpNat kicMsp"
            >
              <h2>
                Dreamy Ink Textures 2
              </h2>
              <p>
                Alcohol Ink Texture Pack
              </p>
            </div>
          </div>
          <div
            className="ListItemFooter sc-ifAKCX gYZfIE"
          >
            <span />
            <p>
              $
              15
            </p>
            <test-file-stub />
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
